console, dashboard: Add persistent, automatically loaded commands history.

cryptomental opened the pull request:
## Overview
Add persistent automatically loaded history file for repl console and Embark dashboard.

Default location of the history file is stored in DEFAULT_CMD_HISTORY_PATH pointing to DAPP_PATH/.embark/cmd_history.

The history is automatically saved and loaded on startup. Repl console has added parametrized .history command that shows last ten commands by default.

Refs: https://github.com/embark-framework/embark/issues/939

### Questions

### Review

### Cool Spaceship Picture
![animated-spaceship-image-0031](https://user-images.githubusercontent.com/20819151/46604802-c5db0c00-caf7-11e8-8f65-bcfdba6e508e.gif)

jrainville commented on cmd/dashboard/command_history.js:
```diff
@@ -24,6 +29,19 @@ class CommandHistory {
+  }
+
+  saveToHistoryFile(cmd) {
+    fs.appendFileSync(this.cmdHistoryFile, cmd + "\n");
```
I think we should limit the number of commands we store here, as it could grow to infinity and also, because getting more than 20 commands in history is useless.

jrainville commented on cmd/dashboard/repl.js:
```diff
@@ -28,6 +34,14 @@ class REPL {
+        .split('\n')
+        .reverse()
+        .forEach((cmd) => { this.replServer.history.push(cmd); })
+    }
```
We should probably use an event to ask the command history module for the command history instead of reading it here.

cryptomental replied:
@jrainville I implemented event-based history for repl and forgot to push during rebase (actually wipe was caused by reset --hard HEAD).... I will have to re-implement this tonight.

jrainville commented on cmd/dashboard/repl.js:
```diff
@@ -36,6 +50,18 @@ class REPL {
+                      .join('\n'));
+        this.displayPrompt();
+      }
+    });
```
You should use the plugin reference (https://embark.status.im/docs/plugin_reference.html#embark-registerConsoleCommand-callback-options)  so that it can be available both in the dashboard and in the console. Also, put the event in the command_history module.

cryptomental replied:
Hi @jrainville thank you for the review. I will start working on this today.

cryptomental commented:
@jrainville I updated the implementation to the improved version and amended the PR.

A few highlights:

* History size is now restricted to process.env.DEFAULT_CMD_HISTORY_SIZE which is set to 20 commands.
* All logic was moved to console module and new registered 'history' command. Repl module uses the event to get commands history
* I also introduced an optional parameter to 'history' that allows to limit output to n latest commands
* Improved error handling

 Example how it works now:

```
Embark (development) > 1
1
Embark (development) > 2
2
Embark (development) > 3
3
Embark (development) > 4
4
Embark (development) > 5
5
Embark (development) > 6
6
Embark (development) > 7
7
Embark (development) > 8
8
Embark (development) > 9
9
Embark (development) > 
(To exit, press ^C again or type .exit)
Embark (development) > 
embark_demo $ ../bin/embark console
Starting Blockchain node in another process
IPFS node not found, attempting to start own node
Starting ipfs process
Blockchain node is ready
ipfs process started
loading solc compiler..
IPFS node detected..
compiling solidity contracts...
deploying contracts
SimpleStorage already deployed at 0x8B326CB9B1da30DeAC147481ef44f05B06e23427
finished deploying contracts
Finished deploying
running webpack with 'development' config...
writing file dist/js/dapp.js
writing file .embark/stats.report
writing file .embark/stats.json
writing file dist/index.html
finished building
Embark (development) > history
9
8
7
6
5
4
3
2
1
Embark (development) > history 3
9
8
7
Embark (development) > history a
Thrown: Invalid argument. Please provide an integer.
Embark (development) > 
(To exit, press ^C again or type .exit)
Embark (development) > 
embark_demo $ cat .embark/cmd_history 
9
8
7
6
5
4
3
2
1
```
I will double check Travis and hopefully the PR can be merged soon.

jrainville commented on cmd/dashboard/command_history.js:
```diff
@@ -24,6 +29,16 @@ class CommandHistory {
+        .reverse()
+        .forEach((cmd) => { this.addCommand(cmd); })
+    }
+  }
```
You can remove this as the console module is the one to handle the file.

cryptomental replied:
Hi @jrainville. That's a tricky one. I was debugging this for a while since the internal dashboard history traversal triggered by up/down keys was not showing up the items and the reason is that the dashboard initialization happens before console service is started. 

To be more specific,

https://github.com/embark-framework/embark/blob/develop/cmd/cmd_controller.js#L109

and console service is started a bit below

https://github.com/embark-framework/embark/blob/develop/cmd/cmd_controller.js#L137

instead of implementing lazy initialization of the history, I decided to leave it in the current state as it  does not seem to do any harm to just pre-initialize dashboard internal history cache and the history is available immediately. No reading files is needed at any point later. Repl console and 'history' command in the console and dashboard are fully event based. I hope that is fine for you. If not we have to figure out another solution.

Btw. Travis & Appveyor are green again! :)

cryptomental commented:
Great! Thanks a lot.

iurimatias commented:
Hi @cryptomental !

I'm getting the following behaviour:

> `embark console`
1
2
3

then ctrl+c and again:

> `embark console`

<up key> => shows "1", expected would be "3"
<up key> => shows "2", expected would be "2"
<up key> => shows "3", expected would be "1"

cryptomental commented:
> Hi @cryptomental !
> 
> I'm getting the following behaviour:
> 
> > `embark console`
> > 1
> > 2
> > 3
> 
> then ctrl+c and again:
> 
> > `embark console`
> 
> => shows "1", expected would be "3"
> => shows "2", expected would be "2"
> => shows "3", expected would be "1"

Hi @iurimatias thanks for noticing! It was caused by array reverse during repl history update. I fixed this and rebased the branch against latest develop, please check again and I hope it is ready for merge .

PascalPrecht commented:
@cryptomental Hey there!

Quick question: is the second commit to make the tests work with the first commit? If so, could we put them into a single one, so we don't introduce a broken state?

cryptomental commented:
Hi @PascalPrecht no problem with that, I will squash the commits asap.

cryptomental commented:
@PascalPrecht commits squashed, tested locally again, works as expected, Travis green. I would be grateful if you had a look again.

cryptomental commented:
AppVeyor green, too. All checks passed.

PascalPrecht commented:
Hey @cryptomental sorry we're quite busy in preps for #CryptoLife. However, this is very likely going to land this week.

cryptomental commented:
Ok @PascalPrecht thank you for the fast reply. Have fun there @CryptoLife ;)
